src/dataframe/dataframe.jl

```diff
@@ -498,51 +498,97 @@ function Base.setindex!(df::DataFrame, v::Any, row_ind::Integer, col_ind::Column
-    if view(_names(df), idxs) != _names(new_df)
-        Base.depwarn("in the future column names in source and target will have to match", :setindex!)
+for T1 in (:AbstractVector, :Not, :Colon),
+    T2 in (:AbstractVector, :Regex, :Not, :Between, :All, :Colon)
+    @eval function Base.setindex!(df::DataFrame,
+                                  new_df::AbstractDataFrame,
+                                  row_inds::$T1,
+                                  col_inds::$T2)
+        idxs = index(df)[col_inds]
+        for (j, col) in enumerate(idxs)
+            df[row_inds, col] = new_df[!, j]
+        end
+        if view(_names(df), idxs) != _names(new_df)
+            Base.depwarn("in the future column names in source and target will have to match", :setindex!)
+        end
+        return df
     end
-    for (j, col) in enumerate(idxs)
-        df[row_inds, col] = new_df[!, j]
+end
+
+for T in (:AbstractVector, :Regex, :Not, :Between, :All, :Colon)
+    @eval function Base.setindex!(df::DataFrame,
+                                  new_df::AbstractDataFrame,
+                                  row_inds::typeof(!),
+                                  col_inds::$T)
+        idxs = index(df)[col_inds]
+        if view(_names(df), idxs) != _names(new_df)
+            throw(ArgumentError("Column names in source and target data frames do not match"))
+        end
+        for (j, col) in enumerate(idxs)
+            # make sure we make a copy on assignment
+            df[!, col] = new_df[:, j]
+        end
+        return df
     end
-    return df
 end
 
 # df[MultiRowIndex, MultiColumnIndex] = AbstractMatrix
-function Base.setindex!(df::DataFrame,
-                        mx::AbstractMatrix,
-                        row_inds::Union{AbstractVector, Not}, # add Colon after deprecation
-                        col_inds::Union{AbstractVector, Regex, Not, Between, All}) # add Colon after deprecation
-    idxs = index(df)[col_inds]
-    if size(mx, 2) != length(idxs)
-        throw(DimensionMismatch("number of selected columns ($(length(idxs))) and number of columns in" *
-                                " matrix ($(size(mx, 2))) do not match"))
-    end
-    for (j, col) in enumerate(idxs)
-        df[row_inds, col] = view(mx, :, j)
+for T1 in (:AbstractVector, :Not, :Colon, :(typeof(!))),
+    T2 in (:AbstractVector, :Regex, :Not, :Between, :All, :Colon)
+    @eval function Base.setindex!(df::DataFrame,
+                                  mx::AbstractMatrix,
+                                  row_inds::$T1,
+                                  col_inds::$T2)
+        idxs = index(df)[col_inds]
+        if size(mx, 2) != length(idxs)
+            throw(DimensionMismatch("number of selected columns ($(length(idxs))) and number of columns in" *
+                                    " matrix ($(size(mx, 2))) do not match"))
+        end
+        for (j, col) in enumerate(idxs)
+            df[row_inds, col] = row_inds isa typeof(!) ? mx[:, j] : view(mx, :, j)
```

Comment:
Are you sure this is better than this? AFAIK the compiler is quite smart with `===`.
```suggestion
            df[row_inds, col] = (row_inds === !) ? mx[:, j] : view(mx, :, j)
```

Reply:
It is I just did want to avoid this as `(! === row_inds)` is a syntax error. But I am OK with the change